routing/cross_mwm_index_graph.cpp

```diff
@@ -41,45 +80,106 @@ bool CrossMwmIndexGraph::IsTransition(Segment const & s, bool isOutgoing)
   auto it = m_transitionCache.find(s.GetMwmId());
   if (it == m_transitionCache.cend())
   {
-    platform::CountryFile const & countryFile = m_numMwmIds->GetFile(s.GetMwmId());
-    TRoutingMappingPtr mappingPtr = m_indexManager.GetMappingByName(countryFile.GetName());
-    CHECK(mappingPtr, ("countryFile:", countryFile));
-    mappingPtr->LoadCrossContext();
-
-    TransitionSegments transitionSegments;
-    mappingPtr->m_crossContext.ForEachOutgoingNode([&](OutgoingCrossNode const & node)
-    {
-      FillTransitionSegments(mappingPtr->m_segMapping, node.m_nodeId, s.GetMwmId(),
-                             transitionSegments.m_outgoing);
-    });
-    mappingPtr->m_crossContext.ForEachIngoingNode([&](IngoingCrossNode const & node)
-    {
-      FillTransitionSegments(mappingPtr->m_segMapping, node.m_nodeId, s.GetMwmId(),
-                             transitionSegments.m_ingoing);
-    });
-    auto const p = m_transitionCache.emplace(s.GetMwmId(), transitionSegments);
-    it = p.first;
-    CHECK(p.second, ("Mwm num id:", s.GetMwmId(), "has been inserted before. countryFile:",
-                     countryFile));
+    FillWholeMwmTransitionSegments(s.GetMwmId());
+    it = m_transitionCache.find(s.GetMwmId());
   }
+  CHECK(it != m_transitionCache.cend(),
+        ("Mwm ", s.GetMwmId(), "has not been downloaded. s:", s, ". isOutgoing", isOutgoing));
 
   if (isOutgoing)
     return it->second.m_outgoing.count(s) != 0;
   return it->second.m_ingoing.count(s) != 0;
 }
 
-void CrossMwmIndexGraph::GetTwins(Segment const & /* s */, std::vector<Segment> & /* twins */) const
+void CrossMwmIndexGraph::GetTwins(Segment const & s, bool isOutgoing, std::vector<Segment> & twins)
 {
+  CHECK(IsTransition(s, isOutgoing), ("IsTransition(", s, ",", isOutgoing, ") returns false."));
   // @TODO(bykoianko) It's necessary to check if mwm of |s| contains an A* cross mwm section
   // and if so to use it. If not, osrm cross mwm sections should be used.
-  NOTIMPLEMENTED();
+  auto const segMwmMapping = GetRoutingMapping(s.GetMwmId());
+  CHECK(segMwmMapping, ("Mwm", s.GetMwmId(), "has not been downloaded. s:", s, ". isOutgoing", isOutgoing));
+
+  twins.clear();
+  vector<string> const & neighboringMwm = segMwmMapping->m_crossContext.GetNeighboringMwmList();
+
+  for (string const & name : neighboringMwm)
+    FillWholeMwmTransitionSegments(m_numMwmIds->GetId(CountryFile(name)));
+
+  auto it = m_transitionCache.find(s.GetMwmId());
+  CHECK(it != m_transitionCache.cend(), ());
```

Comment:
According to the current code it's always true:
- |s| is a transition segment
- method IsTransition(s, isOutgoing) is called form the first CHECK
- InsertWholeMwmTransitionSegments is called in in this IsTransition() if mwm was not loaded

I don't like that we rely on expression in CHECK(). I'll rewrite the code this way:
```
  if (!IsTransition(s, isOutgoing))
    CHECK(false, ("IsTransition(", s, ",", isOutgoing, ") returns false."));
```